src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs

```diff
@@ -367,475 +326,520 @@ private void AnalyzeNodeForDtdProcessingOverloads(SyntaxNodeAnalysisContext cont
                             {
-                                env.IsDtdProcessingSet = true;
+                                xmlReaderSettingsEnv.IsSecureResolver = true;
                             }
-                        }
-
-                        if (isXmlTextReaderXmlResolverProperty &&
-                            (SyntaxNodeHelper.NodeHasConstantValueNull(rhs, model) ||
-                             SecurityDiagnosticHelpers.IsXmlSecureResolverType(rhsType, xmlTypes)))
-                        {
-                            if (env != null)
+                            else if (SecurityDiagnosticHelpers.IsExpressionEqualsNull(operation.Operand))
                             {
-                                env.IsSecureResolver = true;
+                                xmlReaderSettingsEnv.IsSecureResolver = true;
                             }
                         }
-                        else if (isXmlTextReaderDtdProcessingProperty &&
-                                 !SyntaxNodeHelper.GetSymbol(rhs, model).MatchFieldByName(xmlTypes.DtdProcessing, SecurityMemberNames.Parse))
+                        else if (SecurityDiagnosticHelpers.IsXmlReaderSettingsDtdProcessingProperty(prop.InitializedProperty, _xmlTypes))
                         {
-                            if (env != null)
-                            {
-                                env.IsDtdProcessingDisabled = true;
-                            }
+                            xmlReaderSettingsEnv.IsDtdProcessingDisabled = !SecurityDiagnosticHelpers.IsExpressionEqualsDtdProcessingParse(prop.Value);
                         }
-                        else
+                        else if (SecurityDiagnosticHelpers.IsXmlReaderSettingsMaxCharactersFromEntitiesProperty(prop.InitializedProperty, _xmlTypes))
                         {
-                            // Generate a warning whenever the XmlResolver or DtdProcessing property is set to an insecure value
-                            Diagnostic diag = Diagnostic.Create(
+                            xmlReaderSettingsEnv.IsMaxCharactersFromEntitiesLimited = !SecurityDiagnosticHelpers.IsExpressionEqualsIntZero(prop.Value);
+                        }
+                    }
+                }
+            }
+
+            private void AnalyzeVariableDeclaration(OperationAnalysisContext context)
+            {
+                IVariableDeclaration declare = context.Operation as IVariableDeclaration;
+                AnalyzeObjectCreationInternal(context, declare.Variable, declare.InitialValue);
+            }
+
+            private void AnalyzeXmlResolverPropertyAssignmentForXmlDocument(OperationAnalysisContext context, ISymbol assignedSymbol, IAssignmentExpression expression)
+            {
+                XmlDocumentEnvironment env = null;
+                bool isSecureResolver = false;
+
+                if (_xmlDocumentEnvironments.ContainsKey(assignedSymbol))
```

Comment:
This `if` block can moved to the bottom of the method call and be merged with `if (env != null)`